### PR TITLE
[User] Sign up

### PR DESCRIPTION
**Pull request by hatcaiday**
https://edu-redmine.framgia.vn/issues/5013

### CODE REVIEW

**Comment by hatcaiday** [comment]
ready

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
+class UsersController < ApplicationController
+  before_action :load_user, only: %i(show)
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: For one action, no need to declare an array
`before_action :load_user, only: :show`

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  private
+
+  def user_params
+    params.require(:user).permit :username, :email, :password, :password_confirmation, 
+      :dob, :address, :phone
+  end
+
+  def correct_user
+    @user = User.find_by id: params[:id]
+    redirect_to root_url unless current_user? @user
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb:32]: Also add a flash message

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  def load_user
+    @user = User.find_by id: params[:id]
+    return if @user
+    redirect_to root_url
+    flash[:danger] = t ".no_such"
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb:38]: add flash before the redirection

File: app/helpers/users_helper.rb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+module UsersHelper
+end
```
Comment (mdmahmudurrahman) [inline app/helpers/users_helper.rb:23]: Please remove this unused file

File: app/models/user.rb [hunk 5]
```diff
@@ -1,4 +1,37 @@
...
+
+    def new_token
+      SecureRandom.urlsafe_base64
+    end
+  end
+  
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
+    return false if digest.nil?
+    BCrypt::Password.new(digest).is_password?(token)
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Remove optional ()

File: app/models/user.rb [hunk 6]
```diff
@@ -1,4 +1,37 @@
...
+  end
+  
+  private
+
+  def downcase_email
+  	email.downcase!
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Indentation

File: app/views/shared/_message.html.erb [hunk 7]
```diff
@@ -1,3 +1,3 @@
 <% flash.each do |message_type, message| %>
-  <div class="alert alert - <%message_type%>"><%= message%></div>
+  <div class="alert alert-<%message_type%>"><%= message%></div>
```
Comment (mdmahmudurrahman) [inline app/views/shared/_message.html.erb]: `<% message_type %>`
`<%= message %>`

File: app/views/layouts/_header.html.erb [hunk 8]
```diff
@@ -59,7 +59,7 @@
                 <% end %>
               </li>
               <li>
-                <%= link_to "#" do %>
+                <%= link_to signup_path do %>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_header.html.erb]: User can always see this link(logged in user/logged out user) ?

File: app/views/shared/_error_messages.html.erb [hunk 9]
```diff
@@ -0,0 +1,12 @@
+<% if object.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      <%= t ".the_form_contain" %> <%= pluralize object.errors.count, "error" %>.
+    </div>
+    <ul>
+    <% object.errors.full_messages.each do |msg| %>
```
Comment (mdmahmudurrahman) [inline app/views/shared/_error_messages.html.erb]: As this block is nested inside an ul, add 2 space indentation

File: app/views/users/new.html.erb [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+      <h2 class="center-parent"><%= t ".sign_up_title"%></h2>
+      <%= form_for @user, url: signup_path do |f| %>
+        <%= render "shared/error_messages", object: f.object%>
+        <%= f.text_field :username, placeholder: t(".enter_name"), class: "form-control" %>
+        <%= f.email_field :email, placeholder: t(".enter_email"), class: "form-control" %>
+        <%= f.password_field :password, placeholder: t(".enter_pass"), class: "form-control" %>
+        <%= f.password_field :password_confirmation, placeholder: t(".enter_pass"), class: "form-control" %>
+        <div class="form-group">
+          <%= f.label :birthday, class: "form-label" %> 
+          <%= f.date_select :dob, {order: [:month, :day, :year], start_year: Date.today.year - 10, end_year: Date.today.year - 100}, {required: true} %>
```
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: You can write a helper method to define the `start_year` and `end_year`

**Comment by mdmahmudurrahman** [comment]
commented

File: config/routes.rb [hunk 11]
```diff
@@ -1,6 +1,10 @@
...
-  get "/about", to: "static_pages#about"
-  get "/help", to: "static_pages#help"
-  get "/contact", to:"static_pages#contact" 
+  scope "(:locale)", locale: /en|vi/ do
+    root "static_pages#home"
+    get "/help", to: "static_pages#help", as: :help
+    get "/contact", to: "static_pages#contact", as: :contact
+    get "/signup", to: "users#new"
+    post "/signup", to: "users#create"
+    resources :users
```
Comment (mdmahmudurrahman) [inline config/routes.rb]: Specify actions here using `only`

**Comment by hatcaiday** [comment]
ready

File: app/helpers/users_helper.rb [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  def current_user
+    if session[:user_id]
+      @current_user ||= User.find_by id: session[:user_id]
+    end
+  end
+
+  def signup?
+    !current_user.nil?
```
Comment (huongnguyenmta) [inline app/helpers/users_helper.rb]: => `current_user.present?`

File: app/helpers/users_helper.rb [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  def start_year
+    return Date.today.year - 10
```
Comment (huongnguyenmta) [inline app/helpers/users_helper.rb]: `10, 100, ...` move to settings file

**Comment by huongnguyenmta** [comment]
commented

**Comment by hatcaiday** [comment]
ready

File: app/helpers/users_helper.rb [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+    session[:user_id] = user.id  	
+  end
+
+  def current_user
+    if session[:user_id]
+      @current_user ||= User.find_by id: session[:user_id]
+    end
+  end
+
+  def signup?
```
Comment (mdmahmudurrahman) [inline app/helpers/users_helper.rb]: As this method is checking the current user from the session,
please rename this method to indicate whether a user is logged in/not logged in.

**Comment by mdmahmudurrahman** [comment]
(commented)

**Comment by hatcaiday** [comment]
ready

File: app/views/layouts/_header.html.erb [hunk 15]
```diff
@@ -54,13 +54,19 @@
...
-                <%= link_to "#" do %>
+                <%= link_to user_path do %>
                   <i class="fa fa-user"></i><%= t ".user" %>
                 <% end %>
               </li>
               <li>
-                <%= link_to "#" do %>
-                  <i class="fa fa-lock"></i><%= t ".login" %>
+                <% if logged_in? %>
+                  <%= link_to root_path do %>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_header.html.erb:63]: Hope you change it in the next pr :))

**Comment by mdmahmudurrahman** [comment]
(merged)
CC: [To:1818111] Nguyen Thi Huong (94)

**Pull request merged**